### PR TITLE
use async/await keywords when instrumenting asynchronous methods

### PR DESCRIPTION
**Pull request by lucaspimentel**
The initial goal of this PR was to fix an issue with the Web API 2 integration. By using `Task.ContinueWith()`, we were wrapping unhandled exceptions in an `AggregateException`, changing application behavior.

I started refactoring only the `Span.Trace()` extension method to use using `async`/`await` instead of `Task.ContinueWith()`, but ultimately removed that extension method and inlined usages, which in most cases is a simple `try`/`catch`. This results in cleaner code that doesn't wrap exceptions and ensures spans are not disposed before the `Task` finishes.

While debugging this, I also fixed other small issues I found:

- All integrations
  - use `async`/`await` instead of `Task.ContinueWith()`
  - add calls to `Task.ConfigureAwait(false)` where missing
  - renamed `@this` arguments for consistency
- C++ profiler
  - skip more assemblies earlier in the process, don't even look into instrumenting them (`Datadog.Trace`, `Datadog.Trace.ClrProfiler.Managed`, and `Sigil.Emit.DynamicAssembly`)
  - logs: fixed added missing space and removed redundant ones
- Elasticsearch integration
  - fix getting http method
  - fix return type of `CallElasticsearchAsync()` in emitted IL (caught by Sigil)
- misc code clean up (mostly formatting)

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs [hunk 1]
```diff
@@ -69,44 +69,33 @@ public static object Trace(this Span span, Func<object> func, Action<Exception>
...
-
-                if (t.IsCompleted)
-                {
-                    onComplete(null);
-                }
-            });
+                await task.ConfigureAwait(false);
+                onComplete(null);
+            }
+            catch (Exception ex)
```
Comment (bmermet) [inline src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: Nit: would it make sense to use an exception filter as described in this article https://www.thomaslevesque.com/2015/06/21/exception-filters-in-c-6/ to avoid unwinding the stack? In the general case it wouldn't change much, but it would make crash dumps more useful when tracing is enabled.

File: src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs [hunk 2]
```diff
@@ -42,17 +42,17 @@ public static object Trace(this Span span, Func<object> func, Action<Exception>
                     var genericArgs = GetGenericTypeArguments(typeof(Task<>), typ);
                     if (genericArgs.Length == 1)
                     {
-                        if (!_traceTaskMethods.TryGetValue(genericArgs[0], out var traceTask))
+                        if (!_traceTaskMethods.TryGetValue(genericArgs[0], out var traceTaskAsync))
```
Comment (bmermet) [inline src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: It would be nice to have a unit test checking this scenario.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs [hunk 3]
```diff
@@ -24,36 +26,46 @@ public static class AspNetWebApi2Integration
...
-                                            if (e != null)
-                                            {
-                                                scope.Span.SetException(e);
-                                            }
-
-                                            // some fields aren't set till after execution, so repopulate anything missing
-                                            UpdateSpan(controllerContext, scope.Span);
-                                            scope.Span.Finish();
-                                        });
+                try
```
Comment (bmermet) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs]: What prevents using `scope.Span.Trace` here?

File: src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs (hunk 1 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: We're not filtering the exception at all (we want to catch any exception), so an exception filter doesn't make sense here.

Also, I think that article is wrong. Using `throw` (instead of `throw ex`) maintains the exception's original stacktrace. See this other blog post: https://scottdorman.github.io/2007/08/20/Difference-between-quotthrowquot-and-quotthrow-exquot-in-.NET/

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs (hunk 3 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs]: Using `Task.ContinueWith()` wraps exceptions inside an `AggregateException`, changing the behavior of the profiled application. This PR fixes an issue that Kibo is having where they handle a specific exception type and return a 404 http status code, which the front end is expecting. When they enable our tracer, the exception type is wrong and their app returns a 500 instead.

Using `async`/`await` results in cleaner code and does not wrap the exception thrown.

File: src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs (hunk 2 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: Excellent advice. I added unit tests and found a few issues with `Task<T>` here and fixed them. Thanks!

File: src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs [hunk 4]
```diff
@@ -12,7 +9,8 @@ namespace Datadog.Trace.ClrProfiler
     /// </summary>
     public static class Extensions
     {
-        private static ConcurrentDictionary<Type, Func<Span, Task, Action<Exception>, Task>> _traceTaskMethods = new ConcurrentDictionary<Type, Func<Span, Task, Action<Exception>, Task>>();
+        private static readonly ConcurrentDictionary<Type, Func<Task, Action<Exception>, Task>> TraceTaskMethods =
```
Comment (bmermet) [inline src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: Why did you change `TraceTaskMethods`'s casing? `_traceTaskMethods` seemed more consistent to me for a private variable.

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: I changed the casing because I changed the field to `static readonly`.

[Official guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-fields) don't cover internal or private fields, but I usually follow Resharper's recommendation to use lower case for private instance members and upper case for `static readonly` or `const` members.

Comment (bmermet) [reply src/Datadog.Trace.ClrProfiler.Managed/Extensions.cs]: I find it not intuitive, but Resharper must be right 😄 
Does our style checker accept both `_traceTaskMethods` and `TraceTaskMethods`? It would be nice have only one pass the build so that the code stays consistent.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs (hunk 3 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs]: Changing extension method `Span.Trace()` to use `async`/`await` was only temporary. I was going to remove `Span.Trace()` entirely in a separate PR, but I just went ahead and did it here instead (see new commits).

That method was too complicated with little benefit. It checked for return types and had different behavior for `Task`, `Task<T>`, and anything else. It also had another problem: if the result was a `Task` or `Task<T>`, the `Scope` could be disposed before the task was completed.
